File: huffman.py
# Example Huffman coding implementation
# Distributions are represented as dictionaries of { 'symbol': probability }
# Codes are dictionaries too: { 'symbol': 'codeword' }

def huffman(p):
    '''Return a Huffman code for an ensemble with distribution p.'''
    #assert(sum(p.values()) == 1.0) # Ensure probabilities sum to 1

    # Base case of only two symbols, assign 0 or 1 arbitrarily
    if(len(p) == 2):
        return dict(zip(p.keys(), ['0', '1']))

    # Create a new distribution by merging lowest prob. pair
    p_prime = p.copy()
    a1, a2 = lowest_prob_pair(p)
    p1, p2 = p_prime.pop(a1), p_prime.pop(a2)
    p_prime[a1 + a2] = p1 + p2

    # Recurse and construct code on new distribution
    c = huffman(p_prime)
    ca1a2 = c.pop(a1 + a2)
    c[a1], c[a2] = ca1a2 + '0', ca1a2 + '1'

    return c

def lowest_prob_pair(p):
    '''Return pair of symbols from distribution p with lowest probabilities.'''
    assert(len(p) >= 2) # Ensure there are at least 2 symbols in the dist.

    sorted_p = sorted(p.items(), key=lambda (i,pi): pi)
    return sorted_p[0][0], sorted_p[1][0]

def create_prob_dist(text):
    prob_dist = {}
    counts = {}
    total = 0
    for s in text + "!":
        if s in counts.keys():
            counts[s] = counts[s] + 1
            total = total + 1
        else:
            counts[s] = 1
            total = total + 1
    for k in counts.keys():
        prob_dist[k] = round(float(counts[k]) / float(total), 5)
    
    summer = 0.0
    summer = sum(prob_dist.values())
    if summer != 1.0:
        diff = 1.0 - summer
        #print "Difference " + str(diff)
        use_key = prob_dist.keys()[0]
        #print "Key " + use_key
        prob_dist[use_key] = round(prob_dist[use_key] + diff, 5)
    return prob_dist

def hc_encode(text):
    prob_dist = create_prob_dist(text)
    #print prob_dist
    huffman_tree = huffman(prob_dist)
    #print huffman_tree
    code_to = ""
    for s in text:
        code_to = code_to + huffman_tree[s]
    code_to = code_to# + huffman_tree["!"]
    return code_to 
        

def demo_dickens():
    txt = ""
    with open('dickens') as f:
        txt = f.read(200)
    print "Text contains" + str(len(txt)) + " characters"
    xx = hc_encode(txt)
    print "Resulting code: " + str(xx)
    #evaluate
    baits = xx.__len__() / 8
    print baits
    

# Example execution
#ex1 = { 'a': 0.5, 'b': 0.25, 'c': 0.25 }
#a = huffman(ex1) # => {'a': '0', 'c': '10', 'b': '11'}
#print a
#ex2 = { 'a': 0.25, 'b': 0.25, 'c': 0.2, 'd': 0.15, 'e': 0.15 }
#a = huffman(ex2)  # => {'a': '01', 'c': '00', 'b': '10', 'e': '110', 'd': '111'}
#print a

#print hc_encode("abaaaabacbbb")

#print hc_encode("aeaeouuiiiuiui")
#demo_dickens()

File: src/Huffman.java
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.stream.Collectors;


/**
 * Example Huffman coding implementation
 * Distributions are represented as dictionaries of { 'symbol': probability }
 * Codes are dictionaries too: { 'symbol': 'codeword' }
 */

public class Huffman {
	
	public Map<String, String> huffmanTree;
	
	public static void main(String[] args) {
		System.out.println(new Huffman().hcEncode("aeaeouuiiiuiui"));
	}

	
	public Huffman() {
		
	}
	
	public String hcEncode(String text) {
	    Map<String, Double> prob_dist = createProbDist(text);
	    
	    huffmanTree = huffman(prob_dist);
	    
	    String code_to = "";
	    for (int i = 0; i < text.length(); i++) {
	    	String s = Character.toString(text.charAt(i));
	    	code_to += huffmanTree.get(s);
	    }
	    
	    if (!text.equals(decode(code_to))) {
	    	throw new RuntimeException("Decoded string mismatch!!");
	    }
	    
	    return code_to;
	}
	
	public static Map<String, Double> createProbDist(String text) {
	    Map<String, Double> prob_dist = new HashMap<>();
	    Map<String, Integer> counts = new HashMap<>();
	    int total = 0;
	    for (int i = 0; i < text.length(); i++) {
	    	String s = Character.toString(text.charAt(i));
	    
	        if (counts.containsKey(s)) {
	            counts.put(s, counts.get(s) + 1);
	        } else {
	            counts.put(s, 1);
	        }
	        total += 1;
	    }
	    
	    for (String k : counts.keySet()) {
	        prob_dist.put(k, (double)Math.round(((double)(counts.get(k)) / (double)(total))*100000) /100000);
	    }
	    
	    double summer = 0.0;
	    summer = sumVals(prob_dist.values());
	    if (summer != 1.0) {
	        double diff = 1.0 - summer;
	        String use_key = prob_dist.keySet().iterator().next();
	        prob_dist.put(use_key, ((double)Math.round((prob_dist.get(use_key) + diff) * 10000)/10000));
	    }
	
	    return prob_dist;
	}
	
	/*Return pair of symbols from distribution p with lowest probabilities.*/
	public static double sumVals(Collection<Double> values) {
		double summer = 0.0;
		for (Double val : values) {
			summer += val.doubleValue();
		}
		return summer;
	}

	public static String[] lowestProbPair(Map<String, Double> p) {
	    String[] pair = new String[2];
	    assert(p.keySet().size() >= 2); // Ensure there are at least 2 symbols in the dist.
	    
	    Map<String, Double> treeMap = sortByValue(p); 
	    Iterator<Entry<String, Double>> it = treeMap.entrySet().iterator();
	    if(it.hasNext()) {
	    	pair[0] = it.next().getKey();
	    }
	    if(it.hasNext()) {
	    	pair[1] = it.next().getKey();
	    }		
	    return pair;
	}

//pretty print a map
public static <K, V> void printMap(Map<K, V> map) {
    for (Map.Entry<K, V> entry : map.entrySet()) {
        System.out.println("Key : " + entry.getKey()
			+ " Value : " + entry.getValue());
    }
}
	
	/**
	 * Return a Huffman code for an ensemble with distribution p.
	 * */
	public static Map<String, String> huffman( Map<String, Double> p) {
	    // Base case of only two symbols, assign 0 or 1 arbitrarily
	    if (p.keySet().size() == 2) {
	    	HashMap<String, String> hm = new HashMap<>();
	    	String[] keys = new String[2];
	    	p.keySet().toArray(keys);
	    	hm.put(keys[0], "0");
	    	hm.put(keys[1], "1");
	    	return hm;
	    }
	    
	    // Create a new distribution by merging lowest prob. pair
	    Map<String, Double> p_prime = new HashMap<>(p);
	    String[] lowestPair = lowestProbPair(p);
	    Double p1 = p_prime.remove(lowestPair[0]);
	    Double p2 = p_prime.remove(lowestPair[1]);
	    
	    String united = lowestPair[0] + lowestPair[1];
	    
	    p_prime.put(united, p1 + p2);
	    

	    // Recurse and construct code on new distribution
	    Map<String, String> c = huffman(p_prime);
	    
	    String ca1a2 = c.remove(united);
	    c.put(lowestPair[0], ca1a2 + '0');
	    c.put(lowestPair[1], ca1a2 + '1');

	    return c;
	}
	
	/**
	 * Taken from https://www.mkyong.com/java/how-to-sort-a-map-in-java/
	 * and modified for <String, Double>
	 */
	private static Map<String, Double> sortByValue(Map<String, Double> p) {

        // 1. Convert Map to List of Map
        List<Map.Entry<String, Double>> list =
                new LinkedList<Map.Entry<String, Double>>(p.entrySet());

        // 2. Sort list with Collections.sort(), provide a custom Comparator
        //    Try switch the o1 o2 position for a different order
        Collections.sort(list, new Comparator<Map.Entry<String, Double>>() {
            public int compare(Map.Entry<String, Double> o1,
                               Map.Entry<String, Double> o2) {
                return (o1.getValue()).compareTo(o2.getValue());
            }
        });

        // 3. Loop the sorted list and put it into a new insertion order Map LinkedHashMap
        Map<String, Double> sortedMap = new LinkedHashMap<String, Double>();
        for (Map.Entry<String, Double> entry : list) {
            sortedMap.put(entry.getKey(), entry.getValue());
        }

        return sortedMap;
    }
	
	public String decode(String encoded) {
		StringBuilder decoded = new StringBuilder("");
		
		int c = 0;
		StringBuilder reading = new StringBuilder();
		while (c < encoded.length()) {
			reading.append(encoded.charAt(c++));
			String currentRead = reading.toString();
			if (huffmanTree.containsValue(currentRead )) {
				// Find the key which gave the following bitstring
				List<Entry<String, String>> goodValues = huffmanTree
					    .entrySet()
					    .stream()
					    .filter(x -> x.getValue().equals(currentRead ))
					    .collect(Collectors.toCollection(ArrayList::new));
				if (goodValues.size() > 1 || goodValues.size() == 0) {
					throw new RuntimeException("Problem when decoding HC");
				}
				
				decoded.append(goodValues.get(0).getKey());
				reading = new StringBuilder();
			}
			
		}
		
		return decoded.toString();
	}
	
}
